crates/core/tedge_api/src/workflow.rs

```diff
@@ -0,0 +1,586 @@
+        match &state.script {
+            Some(script) => OperationAction::Script(script.to_owned()),
+            None => match &state.owner {
+                Some(owner) if owner == "tedge" => OperationAction::BuiltIn,
+                Some(owner) => OperationAction::Delegate(owner.to_owned()),
+                None => match &state.next[..] {
+                    [] => OperationAction::Clear,
+                    [next] => OperationAction::MoveTo(next.to_owned()),
+                    _ => OperationAction::Delegate("unknown".to_string()),
+                },
+            },
+        }
+    }
+}
+
+/// Generic command state that can be used to manipulate any type of command payload.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct GenericCommandState {
+    pub topic: String,
+    pub status: String,
+    pub payload: Value,
+}
+
+impl GenericCommandState {
+    /// Extract a command state from a json payload
+    pub fn from_slice(topic: String, bytes: &[u8]) -> Result<Self, WorkflowExecutionError> {
+        let json: Value = serde_json::from_slice(bytes)?;
+        let status = GenericCommandState::extract_text_property(&json, "status")
+            .ok_or(WorkflowExecutionError::MissingStatus)?;
+        Ok(GenericCommandState {
+            topic,
+            status,
+            payload: json,
+        })
+    }
+
+    /// Serialize the command state as a json payload
+    pub fn to_json_string(mut self) -> String {
+        GenericCommandState::inject_text_property(&mut self.payload, "status", &self.status);
+        self.payload.to_string()
+    }
+
+    /// Inject a json payload into this one
+    pub fn update_from_json(mut self, json: Value) -> Self {
+        if let (Some(values), Some(new_values)) = (self.payload.as_object_mut(), json.as_object()) {
+            for (k, v) in new_values {
+                values.insert(k.to_string(), v.clone());
+            }
+        }
+        match GenericCommandState::extract_text_property(&self.payload, "status") {
+            None => self.fail_with("Unknown status".to_string()),
+            Some(status) => GenericCommandState { status, ..self },
+        }
+    }
+
+    /// Update the command state with the outcome of a script
+    pub fn update_with_script_output(
+        self,
+        script: String,
+        output: std::io::Result<std::process::Output>,
+    ) -> Self {
+        match output {
+            Ok(output) => {
+                if output.status.success() {
+                    match String::from_utf8(output.stdout) {
+                        Ok(stdout) => match serde_json::from_str(&stdout) {
+                            Ok(json) => self.update_from_json(json),
+                            Err(err) => {
+                                let reason =
+                                    format!("Script {script} returned non JSON stdout: {err}");
+                                self.fail_with(reason)
+                            }
```

Comment:
I was under the assumption that the successful exit code is enough for the workflow to move to the next "target state" irrespective of the output format, and if the output ends up being JSON then we append it to the command state message.